teps/0011-redirecting-step-output-streams.md

```diff
@@ -0,0 +1,219 @@
+
+### Option 2: Redirecting to Canonical Conventional Paths
+
+The following new fields will be added to the `Step` struct:
+
+```go
+type Step struct {
+    ...
+    // Whether to capture the stdout stream. If set, the stream will be duplicated to `/tekton/steps/<step_index>/stdout`.
+    // +optional
+    Stdout bool `json:"stdout,omitempty"`
+    // Whether to capture the stderr stream. If set, the stream will be duplicated to `/tekton/steps/<step_index>/stderr`.
+    // +optional
+    Stderr bool `json:"stderr,omitempty"`
+}
+```
+
+Once `Stdout` or `Stderr` is set, the corresponding output stream will be duplicated to both the conventional path indicated above and the standard output stream of the container, so users can still view the output through the Pod log API. Variable substitutions for `$(steps.<name>.stdoutPath)` and `$(steps.<name>.stderrPath)` will be provided if the corresponding field is set to grant users easy access to the conventional paths.
+
+#### Example Usage
+
+Redirecting stdout of `boskosctl` to `jq` and publish the resulting `project-id` as a Task result:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: boskos-acquire
+spec:
+  results:
+  - name: project-id
+  steps:
+  - name: boskosctl
+    image: gcr.io/k8s-staging-boskos/boskosctl
+    args:
+    - acquire
+    - --server-url=http://boskos.test-pods.svc.cluster.local
+    - --owner-name=christie-test-boskos
+    - --type=gke-project
+    - --state=free
+    - --target-state=busy
+    stdout: true
+  - name: parse-project-id
+    image: imega/jq
+    args:
+    - -r
+    - .name
+    - $(steps.boskosctl.stdoutPath)
+    stdout: true
+  - name: copy-result
+    image: alpine
+    args:
+    - cp
+    - $(steps.parse-project-id.stdoutPath)
+    - $(results.project-id.path)
+```
+
+#### Risks and Mitigations
+
+* Task authors cannot configure the redirection path, it would take extra steps to implement the following use cases:
+  1. Write stdout of a step into `.docker/config.json` and consume the generated auth configuration in a subsequent step.
+  1. Extract stdout into a Task result.
+  
+  If the step image provides a shell, users can specify the `Script` field to overwrite the image entrypoint to invoke the original entrypoint command and redirect the stdout into the target file. There are two main drawbacks: 1) information encapsulated by the image (e.g., entrypoint) will be leaked into Task specification, creating a tight coupling between the Task and the image; 2) users will lose the ability to see stdout through Pod log API.
```

Comment:
nit: missing `*` bullet point here?

Reply:
No I want to further discuss how to mitigate the problem raised in this bullet. Let me work on the flow a bit.

Reply:
OK, no biggie if markdown fails us here, just wanted to check it wasn't a separate item.